Move dependencies to pyproject toml and create ci/cd install and import test

leifdenby opened the pull request:
The aim of this PR is to move dependencies from `requirements.txt` to `pyproject.toml` (as has been proposed for `v0.2.0` on [the roadmap](https://docs.google.com/document/d/1csgMjyuE1Up2m262e-N7qMPmNJ-zHLguGyrgvnjTuBE/edit)) and discussed on https://github.com/mllam/neural-lam/pull/27#issuecomment-2110001148.

Requirements:
- dependencies should still be installable with `pip` and `pyproject.toml` can be updated by hand
- dependencies can be handle with [pdm](https://pdm-project.org/en/stable/)
- ci/cd test that runs the pip install and runs a test that imports `neural-lam` to ensure that the required packages have correctly been installed.

leifdenby commented:
I will copy here my motivation for allow using of `pdm` from https://github.com/mllam/neural-lam/pull/27#issuecomment-2117603643:

For someone not using `pdm` the process of setting up a new development environment would be:

1. Clone repo `git clone http://github.com/<username>/neural-lam` and change directory `cd neural-lam`
2. Optionally create your virtualenv with your favourite tool, maybe `mkvirtualenv ...` and then `activate ...`
3. Install `neural-lam` editable with pip so that changes you make are reflected in the `site-packages` version (actually this is simply a link) `pip install -e .`
4. Edit and run `neural-lam`: `python -m neural_lam.train_model --config-file config.yaml`
5. Manually install a dependency with `pip` and edit `pyproject.toml` by hand, i.e. `python -m pip install ...` and edit the `pyproject.toml` file
6. Stage, commit and push!

I like using a package manager like [pdm](https://pdm-project.org) (other options are [pipenv](https://pipenv.pypa.io/en/latest/), [poetry](https://python-poetry.org/), [uv](https://github.com/astral-sh/uv)) because it makes the development process easier by 1) automatically updating package versions in `pyproject.toml` to ensure versions work together when I add a new package, 2) allowing me to "dry run" any dependency modification without clobbering my current env (this has hurt me quite a few times with pip) and 3) handling creating and activating virtuals envs

So with `pdm` the development process would be:

1. Clone repo `git clone http://github.com/<username>/neural-lam` and change directory `cd neural-lam`
2. Create virtual env `pdm venv create` and either activating it `pdm venv activate ...` or making it the default when using `pdm run python ...` with `pdm use ...`
3. Install package `pdm install`
4. Add dependency `pdm add ...` (this is where you can do ` pdm add --dry-run ...` to see what would change before you install a package), or remove one with `pdm remove ...`. You can also add `dev` dependencies separately (which won't be included in wheels) with `pdm add --dev ...` or dependency groups with `pdm add --group visualisation matplotlib` for example (if we didn't want visualisation tools installed by default)
5. Edit and run `neural-lam`: `python -m neural_lam.train_model --config-file config.yaml` (if using activated virtualenv) or `pdm run python -m neural_lam.train_model --config-file config.yaml`
6. Stage, commit and push!

sadamov commented:
Since our cluster upgraded to cuda version 12.0 I ran into some issues with pinned dependencies (e.g. pyg-lib=0.2.0 is not avaiable: https://pytorch-geometric.com/whl/torch-2.0.1%2Bcu121.html). When we install dependencies with pdm in the future we should make sure that cuda < and > 12.0 is supported. Unless we expect the user to install their own cuda framework which I think is too limiting. In case it is useful, this is the command I am currently using to install the neural-lam env with cuda support:
```
mamba install pytorch pytorch-lightning torchvision torchaudio pytorch-cuda=12.1 pyg "numpy<2.0.0" cartopy pandas wandb matplotlib scipy shapely networkx cartopy pyproj plotly xarray zarr dask pre-commit geopandas pytest pooch -c pyg -c pytorch -c nvidia
pip install tueplots
```

joeloskarsson commented:
Do you know if there is a way to install the dev dependencies using pip? Or do you have to do this manually now that they are listed under `[tool.pdm.dev-dependencies]`? I tried a few things but could not figure it out

leifdenby commented:
@khintz and I finally got testing on GPUs (using cirun.io running on AWS EC2 with GPU) with both pure pip and pdm based installs working yesterday 🥳 As you will see there are now four different github action workflows for each of the four variations of cpu/gpu and pip/pdm. Doing this work I learnt that:

- `torch` comes as different "variants" for installing for cpu-only support or support for a specific CUDA version, and the default variant that is installed with pip or pdm (when simply doing `pip install torch` or `pdm add torch`) is the latest torch version with latest CUDA version. If one wants specific variant other than default (for example for developing without a GPU) then this can only be achieved by providing an additional pip package index URL when installing with pip (with `--index-url https://download.pytorch.org/whl/cpu` e.g. the command would be `pip instal torch --index-url https://download.pytorch.org/whl/cpu`)
- [pdm does not support "index urls" when adding a package](https://github.com/pdm-project/pdm/discussions/1032), there are workaround where we add index URLs directly to the `pyproject.toml`, but then we would have to add all the variants we think people might use... There are [more complicated solutions](https://github.com/EmbarkStudios/pdm-plugin-torch). I have opted for simply suggesting that people manually install the torch variant they want with `pip` before doing `pdm install` if they don't want the latest torch version with the latest CUDA version.
- because variants can only be installed by the providing this additional index url then
    - I cannot do a pure pdm-based install to get the cpu variant installed, I have to first create a virtual environment, then install torch with torch+cpu with pip and then complete the install with pdm. As long as the version of the torch package installed is adequate then pdm will be happy with it and won't try to install the latest gpu variant by default
    - We can't include a pdm.lock file (which would pin the specific versions of a working install, i.e. be the result of the dependency resolution from the `pyproject.toml` file) because that would refer to a specific torch variant and then we would be locked out of installing other variants with pdm. We could of course in future consider maintaining a collection of lock files (this could be good for productions setup), but I think that is out of scope for now.

In the github actions you will also see that explicitly check that the torch variant has been installed. So I hope this will a solid enough setup going forward.

Anyway, the good news this now all works! I have updated the [README](https://github.com/leifdenby/neural-lam/blob/maint/deps-in-pyproject-toml/README.md#installation) with install instructions and this is all ready for review 😄 And in future we will have both CPU and GPU based testing running in ci/cd.

> Since our cluster upgraded to cuda version 12.0 I ran into some issues with pinned dependencies (e.g. pyg-lib=0.2.0 is not avaiable: https://pytorch-geometric.com/whl/torch-2.0.1%2Bcu121.html). When we install dependencies with pdm in the future we should make sure that cuda < and > 12.0 is supported.

I think this should work with the setup I have implemented now. Could you try with the install instructions I have written? I hope this works... 🤞

leifdenby commented:
> Do you know if there is a way to install the dev dependencies using pip? Or do you have to do this manually now that they are listed under `[tool.pdm.dev-dependencies]`? I tried a few things but could not figure it out

There isn't no, and this is be default I found out. The reason is that in pdm they have taken the policy that "dev"-dependencies shouldn't be part of the package that is built when putting a package on e.g. `pypi`. We could simply choose to not use the pdm special dev-dependencies, and instead use a optional extras group which we could also call `dev`.

The install would then change from

```
pdm install --dev
```

to 

```
pdm instal --group dev
```

And with pip you could then install this with

```
python -m pip install .[dev]
```

I would be ok with this if people prefer this?

joeloskarsson commented:
I see. Yes, I think using an optional extras group is preferable then, since it would work smoothly both wiht pip and pdm.

leifdenby commented:
> I see. Yes, I think using an optional extras group is preferable then, since it would work smoothly with both pip and pdm.

Cool. Yes I think you're right. I would imagine @sadamov, @SimonKamuk and @khintz are happy with this too. If I don't hear any complaints I will change that this morning and update the PR (README and pyproject.toml)

I will take that is your review so far @joeloskarsson 😀

joeloskarsson commented:
I have looked over all the changes and I think it looks good. Want to do a full install on my end and see that everything works, but will wait with that until you've made that final change. I have not looked in detail on any of the github workflows, as I don't know that stuff so well. But I trust others to go over that.

Does this go in before or after #32 ?

leifdenby commented:
> I have looked over all the changes and I think it looks good. Want to do a full install on my end and see that everything works, but will wait with that until you've made that final change. I have not looked in detail on any of the github workflows, as I don't know that stuff so well. But I trust others to go over that.

Great, thanks!


> Does this go in before or after #32 ?

After I would say. The other PR is about moving files around to make neural-lam into a package. Once we've done that I can make sure that package installs and runs correctly for both cpu and gpu support

sadamov reviewed (changes requested):
Looking great! Thanks, really also for the work on the github runners which must have been quite a hassle to set up :rocket: 

I have tried installation with pdm (4m30s) and pip (6m54s) (latest torch version) on my HPC system running CUDA 12.0 and can report that for both envs neuralLAM-training was successful. Tested both on cpu and gpu.

After capping the numpy version this can be merged, from my side. I trust that the `--dev` install will work after the change discussed above.

joeloskarsson commented on pyproject.toml:
```diff
@@ -1,3 +1,39 @@
+]
+requires-python = ">=3.9"
+
+[tool.pdm.dev-dependencies]
+dev = [
+    "pre-commit>=2.15.0",
+    "pytest>=8.2.1",
+    "pooch>=1.8.1",
+]
```
Change to optional dependency group called just `dev`, to allow for installing both using pdm and pip.

leifdenby replied:
All done! :)

joeloskarsson reviewed (changes requested):
Change the numpy version spec and the dev group name, then this is all good to go! 

Something I realized, that's really good, is that we don't actually need any of the optional PyG dependencies atm. So this getting rid of them is great, and also leads to less complications with torch CPU/CUDA versions. We can just re-introduce them later if some will be needed.

sadamov commented on pyproject.toml:
```diff
@@ -1,3 +1,39 @@
+# PEP 621 project metadata
+# See https://www.python.org/dev/peps/pep-0621/
+dependencies = [
+    "numpy>=1.24.2",
```
I apologise, it was a me-problem working in a potentially dirty env with multiple numpy-versions. In https://github.com/mllam/neural-lam/issues/67 we demonstrate that neural-lam is ready for numpy>=2.0.0. No need to cap it!

leifdenby replied:
Thanks for the work on #67 I will "resolve" this discussion :)

leifdenby commented:
Ok @sadamov and @joeloskarsson, I've moved the dev dependencies to the optional dependency groups and updated the ci/cd setups and README to reflect the change.

joeloskarsson commented:
* Now this includes ( a bit hidden) the upgrade to numpy 2.0. Can we make this a bit more transparent? For example by adding this specifically to the changelog, as it now just contains the entry for capping it to <2.0.
* If we now are moving to allow numpy 2.0, should we actually make this `>=2.0.0`, so we do not need to make sure to support both numpy 1 and 2 in the future?

leifdenby commented:
> * Now this includes ( a bit hidden) the upgrade to numpy 2.0. Can we make this a bit more transparent? For example by adding this specifically to the changelog, as it now just contains the entry for capping it to <2.0.
> 
> * If we now are moving to allow numpy 2.0, should we actually make this `>=2.0.0`, so we do not need to make sure to support both numpy 1 and 2 in the future?

Ah! Sorry, that was an oversight. I had somehow convinced myself we now *did* support `numpy>=2.0.0`, but of course #67 capped the version. I'm glad you noticed this. I will make another commit to cap the version as it was before in `requirements.txt`

joeloskarsson commented:
Oh no, sorry for the confusion. We _want_ to remove the cap to <2.0 here, just make it clear that we do that. The code works with numpy 2, see

sadamov reviewed (approved):
I installed the dev envs with both pip and pdm now, it works!
One model training was also succesful with new envs.

I agree to remove the numpy cap and mention it in the changelog of this PR.

Happy merging :tada:

leifdenby commented:
> Oh no, sorry for the confusion. We _want_ to remove the cap to <2.0 here, just make it clear that we do that. The code works with numpy 2, see #67

Ah I see! Ok, so I will include a note in the CHANGELOG to say that we removed the cap to `numpy<2.0.0` because it works. I'll draft that changelog entry now and revert the changes

joeloskarsson reviewed (approved):
Great! Now everything looks good to me.
